glue/core/tests/test_edit_subset_mode.py

```diff
@@ -87,3 +94,81 @@ def test_combines_make_copy(self):
         self.edit_mode.edit_subset = self.data.new_subset()
         self.edit_mode.update(self.data, self.state2)
         assert self.edit_mode.edit_subset.subset_state is not self.state2
+
+
+def test_no_double_messaging():
+
+    # Make sure that when we create a new subset via EditSubsetMode, we don't
+    # get two separate messages for creation and updating, but instead just a
+    # single create message with the right subset state.
+
+    handler = MagicMock()
+
+    app = Application()
+
+    class Listener(HubListener):
+        pass
+
+    listener = Listener()
+
+    app.session.hub.subscribe(listener, SubsetCreateMessage, handler=handler)
+    app.session.hub.subscribe(listener, SubsetUpdateMessage, handler=handler)
+
+    data = Data(x=[1, 2, 3, 4, 5])
+
+    app.data_collection.append(data)
+
+    cmd = ApplySubsetState(data_collection=app.data_collection,
+                           subset_state=data.id['x'] >= 3,
+                           override_mode=ReplaceMode)
+
+    app.session.command_stack.do(cmd)
+
+    assert handler.call_count == 1
+    message = handler.call_args[0][0]
+    assert isinstance(message, SubsetCreateMessage)
+    assert isinstance(message.subset.subset_state, InequalitySubsetState)
+    assert message.subset.subset_state.left is data.id['x']
+    assert message.subset.subset_state.operator is operator.ge
+    assert message.subset.subset_state.right == 3
+
+
+def test_message_once_all_subsets_exist():
+
+    # Make sure that when we create a new subset via EditSubsetMode, we don't
+    # get two separate messages for creation and updating, but instead just a
+    # single create message with the right subset state.
+
+    app = Application()
+
+    class Listener(HubListener):
+        pass
+
+    listener = Listener()
+
+    data1 = Data(x=[1, 2, 3, 4, 5])
+    app.data_collection.append(data1)
+
+    data2 = Data(x=[1, 2, 3, 4, 5])
+    app.data_collection.append(data2)
+
+    count = [0]
+
+    def handler(msg):
+        assert len(data1.subsets) == len(data2.subsets)
```

Comment:
```suggestion
        assert len(data1.subsets) == len(data2.subsets)
        assert data2.subsets[0].subset_state.left is data.id['x']
        assert data2.subsets[0].subset_state.operator is operator.ge
        assert data2.subsets[0].subset_state.right == 3
```
Looking again at the suggested `test_broadcast_to_collection` above, I am wondering if the extra checks could simply be integrated here. Or rather
`assert data2.subsets[0].subset_state.foo == assert data1.subsets[0].subset_state.foo`?

Reply:
The subset state objects might actually be identical and could check with 'is'? (So not bothering to check attributes of the subset states since the subset states are the same)

Reply:
That fails with
```python
>       assert data2.subsets is data1.subsets
E       assert (Subset: Subset 1 (data: ),) is (Subset: Subset 1 (data: ),)
E        +  where (Subset: Subset 1 (data: ),) = Data (label: ).subsets
E        +  and   (Subset: Subset 1 (data: ),) = Data (label: ).subsets
```
(as well as `data2.subsets[0] is data1.subsets[0]`, so apparently it's a copy) but
`assert data2.subsets[0].subset_state is data1.subsets[0].subset_state`
works.

Reply:
Yes sorry I was referring to subset_state